fix: Improve user task assignment handling and `changedAttributes` population

ce-dmelnych opened the pull request:
## Description
TBD

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/usertask/AssignUserTaskTest.java:
```diff
@@ -84,6 +85,7 @@ public void shouldEmitAssignEventsForUserTask() {
                     .hasUserTaskKey(userTaskKey)
                     .hasAction(DEFAULT_ACTION)
                     .hasAssignee("foo")
+                    .hasOnlyChangedAttributes(UserTaskRecord.ASSIGNEE)
```
😍

korthout commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/UserTaskHandler.java:
```diff
@@ -233,6 +233,14 @@ private boolean refersToPreviousVersionRecord(final long key) {
+    } else {
+      entity.setAssignee(record.getValue().getAssignee());
+    }
+  }
```
🔧 This method can be static
🔧 To increase clarity, please consider swapping the parameters. Having the entity that will be changed first is common practice in Java APIs (although not documented for our codebase)

ce-dmelnych replied:
Thank you, Nico, for the feedback!  

We addressed this together by transforming:  
```java
private setAssignee(final Record<UserTaskRecordValue> record, final TaskEntity entity)
```  
into:  
```java
private static String getAssigneeOrNull(final Record<UserTaskRecordValue> record)
```  

This refactor enabled us to reuse the `getAssigneeOrNull` method in both the `updateChangedAttributes` and `createTaskEntity` methods.

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java:
```diff
@@ -373,17 +380,20 @@ public void shouldTriggerAssignmentListenersAfterUserTaskCreationWithDefinedAssi
-            tuple(UserTaskIntent.COMPLETE_TASK_LISTENER, assignee, action),
-            tuple(UserTaskIntent.COMPLETE_TASK_LISTENER, assignee, action),
-            tuple(UserTaskIntent.ASSIGNED, assignee, action));
+            tuple(UserTaskIntent.CREATING, StringUtils.EMPTY, action, List.of()),
+            tuple(UserTaskIntent.CREATED, StringUtils.EMPTY, action, List.of()),
+            tuple(UserTaskIntent.ASSIGNING, assignee, action, List.of(UserTaskRecord.ASSIGNEE)),
+            tuple(UserTaskIntent.COMPLETE_TASK_LISTENER, assignee, action, List.of()),
+            tuple(UserTaskIntent.COMPLETE_TASK_LISTENER, assignee, action, List.of()),
+            tuple(UserTaskIntent.COMPLETE_TASK_LISTENER, assignee, action, List.of()),
+            tuple(UserTaskIntent.ASSIGNED, assignee, action, List.of(UserTaskRecord.ASSIGNEE)));
```
🔧 Would be interesting to see this same assertion for a case with two additional corrected properties

❌ I've quickly tested this and noticed that the corrected attributes don't appear in the changed attributes of the `ASSIGNED` event

ce-dmelnych replied:
Thank you for the suggestion!
I’ve tested this scenario in the `TaskListenerTest#shouldTrackOnlyChangedAttributesDuringUserTaskAssignmentWithMultipleListeners` test case.

This test specifically verifies how multiple listeners handle corrections to user task properties and ensures that only the truly changed attributes are tracked in the `ASSIGNED` event.

korthout reviewed (changes requested):
Nice work @ce-dmelnych 👏 

❌ After some testing, I feel like we're not there yet completely.

- Corrected attributes don't appear in the changed attributes anymore, e.g. correct priority on assigning should have both in changed
- We should only add values to changed attributes, if they actually differ with the previous state. I.e. don't add assignee as changed when assigning to the same assignee a second time.

tsedekey commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/UserTaskHandler.java:
```diff
@@ -233,15 +233,12 @@ private boolean refersToPreviousVersionRecord(final long key) {
-    entity.getChangedAttributes().add("assignee");
+  private void setAssignee(final Record<UserTaskRecordValue> record, final TaskEntity entity) {
     if (ExporterUtil.isEmpty(record.getValue().getAssignee())) {
       entity.setAssignee(null);
```
I was testing your code locally, and it seems that setting the assignee property to null is not being persisted for some reason. The get task service still returns the assignee property as an empty string. I wasn’t able to verify this on ES because I’m not very familiar with ES yet.

ce-dmelnych replied:
Thank you for sharing your feedback, Eddie!

Could you share how you tested the new code from the branch?
Did you use the `Makefile` from the `tasklist` mvn module, or did you follow a different approach?

I'd like to understand your setup so I can try to reproduce the issue as well.

ce-dmelnych commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/usertask/AssignUserTaskTest.java:
```diff
@@ -70,10 +74,17 @@ public void shouldEmitAssigningEventForUserTask() {
+        RecordingExporter.userTaskRecords(UserTaskIntent.ASSIGNED).getFirst().getValue();
+
+    assertThat(List.of(assigningRecordValue, assignedRecordValue))
+        .allSatisfy(
```
I've added a `describedAs` message to clarify that the data consistency between `ASSIGNING`/`CLAIMING` and `ASSIGNED` records is essential, as dependent applications like Tasklist and Operate will rely on it to update user task data internally.

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/processors/UserTaskAssignProcessor.java:
```diff
@@ -54,7 +54,11 @@ public void onCommand(
     final long userTaskKey = command.getKey();
 
-    userTaskRecord.setAssignee(command.getValue().getAssignee());
+    final var newAssignee = command.getValue().getAssignee();
+    if (!userTaskRecord.getAssignee().equals(newAssignee)) {
+      userTaskRecord.setAssignee(newAssignee);
+      userTaskRecord.setAssigneeChanged();
+    }
```
@korthout 

As we are introducing this check to compare the new `assignee` with the persisted user task `assignee` in both `UserTaskAssignProcessor` and `UserTaskClaimProcessor`. This ensures that the `changedAttributes` and `assignee` will be updated only when `assignee` was actually changed.

With this in place, it seems like there’s no need to add an additional check like `modifiedRecord.setDiffAsChangedAttributes(persistedRecord);` in the `UserTaskProcessor#handleCommandProcessing` when no task listeners are defined before finalizing command.

I left the `UserTaskProcessor` class unchanged for now, but we might revisit this approach during the implementation of `updating` listeners.

What do you think about this?

korthout replied:
Should be fine for now I think

ce-dmelnych commented:
Hey @panagiotisgts,

It seems @tsedekey is on a long vacation 👶🏻🍼. Could you, on behalf of the HTO team, review this PR?
This is especially relevant as we modified the [UserTaskHandler](https://github.com/camunda/camunda/pull/26877/files#diff-9b1d49038165041139dbd9c3172f71f3702b14380388f23731fded99dac0a8e6) class in the `exporter` module.  

The key change is that with the `ASSIGNED` user task record, Zeebe now includes the "assignee" entry in the `changedAttributes` list when the `assignee` property is modified. This allows the `UserTaskHandler` to rely solely on `record.getValue().getChangedAttributes()` to apply updates to the `entity` based on the changed attributes.  

Additionally, we've fixed an issue where the `entity.assignee` property was incorrectly assigned an empty string `""` after handling user task assignment, instead of being set to `null`.  

Your review would be greatly appreciated!

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java:
```diff
@@ -1363,6 +1363,116 @@ private void testProvideCorrectedUserTaskDataToSubsequentTaskListener(
+            // initial values that were before task assignment using the corrections.
+            tuple(UserTaskIntent.ASSIGNED, List.of("candidateGroupsList", "dueDate")));
+
+    final var createUserTaskValue =
```
```suggestion
    final var createdUserTaskValue =
```

korthout reviewed (approved):
Thanks @ce-dmelnych 👏 

LGMT 👍

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnUserTaskBehavior.java:
```diff
@@ -330,6 +330,7 @@ public void userTaskCreated(final UserTaskRecord userTaskRecord) {
   public void userTaskAssigning(final UserTaskRecord userTaskRecord, final String assignee) {
     final long userTaskKey = userTaskRecord.getUserTaskKey();
     userTaskRecord.setAssignee(assignee);
+    userTaskRecord.setAssigneeChanged();
```
Currently, we invoke the `BpmnUserTaskBehavior#userTaskAssigning` method only when the user task is configured with a defined (non-empty string) assignee. 

However, to make it safer and consistent with the checks we already have in the `UserTaskAssignProcessor` and `UserTaskClaimProcessor` classes, I've added a similar condition here. This improvement ensures better alignment and might also be helpful if we decide to reuse this method in `UserTaskAssignProcessor` later.
